### PR TITLE
Add coral-dbt testing framework

### PR DESCRIPTION
**Pull request by AliceYeh12**
**Summary**
Set up testing framework in `coral-dbt` to test incremental maintenance Scala code generation logic, which will run on build.

**Changes/Additions**

- Set up Python `unittest` module and add unit tests for simple select and join
- Add scripts to copy and format macros for testing and clean up on tear down
- Modify `build.gradle` to run tests on build
- Refactor macros

**Testing**
Ran build locally in a virtual environment. On build, the following lines are now expected on success:
```
test_join (__main__.TestIncremental) ... ok
test_simple_select_all (__main__.TestIncremental) ... ok

----------------------------------------------------------------------
Ran 2 tests in 0.076s

OK
```

### CODE REVIEW

File: coral-dbt/src/main/resources/tests/test_incremental.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        output = template.render(context)
+
+        return output
+
+class TestIncremental(unittest.TestCase):
+    maxDiff = None
+
+    @classmethod
+    def setUpClass(self):
+        subprocess.run(["chmod", "+x", "./setup_test.sh"])
```
Comment (wmoustafa) [inline coral-dbt/src/main/resources/tests/test_incremental.py:42]: Why is this needed?

**Comment by wmoustafa** [comment]
The build workflow has failed in Github. Could you verify what went wrong?

**Comment by AliceYeh12** [comment]
> The build workflow has failed in Github. Could you verify what went wrong?

@wmoustafa This is the intended behavior at the moment—it is failing on the coral-dbt tests that this framework introduces since the expected values are based on the new incremental logic that [PR #376](https://github.com/linkedin/coral/pull/376) introduces, which this branch has not been rebased on yet.

File: coral-dbt/src/main/resources/tests/setup_test.sh [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+if [ $# -ne 2 ]; then
+    echo "Usage: $0 source_file destination_file"
+    exit 1
+fi
+
+# Copy the file
+cp $1 $2
+
+# Remove copyright and comments (will otherwise be printed) and return call
+sed -i -e '/\/\*/,/\*\//d ; /^--/d ; s/return(\(.*\))/\1/' $2
```
[the remark this answers is not in the document]
Comment (AliceYeh12) [reply coral-dbt/src/main/resources/tests/setup_test.sh:13]: Sure, the first expression matches anything starting with `/*` up to `*/`, which should match the multi-line comment structure (used by the copyright). I'm not too concerned with this accidentally matching something else since anything with this format should be a comment. 

```
/**
 * Copyright 2023 LinkedIn Corporation. All rights reserved.
 * Licensed under the BSD-2 Clause license.
 * See LICENSE in the project root for license information.
 */
```

The second expression matches anything starting with `--`, which is the single line comment structure; again, also not too worried about this unintentionally matching something.

```
-- This is a comment
```

The third expression replaces a `return(foo)` statement with just `foo`, without loss of generality. This should not generate a mismatch either since there should be only one relevant return statement per macro. 

```
// Before
{{ return(foo) }}

// After
{{ foo }}
```

File: coral-dbt/src/main/resources/tests/test_incremental.py (hunk 1 above)
Comment (AliceYeh12) [reply coral-dbt/src/main/resources/tests/test_incremental.py:42]: The command is to make the `setup_test.sh` file executable.

File: coral-dbt/build.gradle [hunk 3]
```diff
@@ -5,4 +5,40 @@ spotless {
...
+  process.waitFor()
+  def errorStream = process.getErrorStream()
+  def buffReader = new BufferedReader(new InputStreamReader(errorStream))
+  def stringBuilder = new StringBuilder()
+  def newLine
+  while ((newLine = buffReader.readLine()) != null) {
+    stringBuilder.append(newLine)
+  }
+  def errorOutput = stringBuilder.toString()
+  if (errorOutput.contains("AssertionError")) {
```
[the remark this answers is not in the document]
Comment (AliceYeh12) [reply coral-dbt/build.gradle:39]: Ah, yes, modified to error based on process's exit code. The error output on success is not too interesting—it would look something like:

```
..
----------------------------------------------------------------------
Ran 2 tests in 0.104s

OK
```
Not sure if this is worth including.

File: coral-dbt/src/main/resources/macros/coral_macros/spark/utils/get_coral_incremental_response.sql [hunk 4]
```diff
@@ -3,10 +3,9 @@
  * Licensed under the BSD-2 Clause license.
  * See LICENSE in the project root for license information.
  */
-{% macro spark__get_coral_incremental_response(sql, table_names) -%}
+{% macro spark__get_coral_incremental_response(sql, url, table_names) -%}
```
[the remark this answers is not in the document]
Comment (AliceYeh12) [reply coral-dbt/src/main/resources/macros/coral_macros/spark/utils/get_coral_incremental_response.sql:6]: No, it is not required. Will move logic over to default.

File: coral-dbt/src/main/resources/macros/coral_macros/spark/materializations/incremental_maintenance.sql [hunk 5]
```diff
@@ -17,47 +17,14 @@
...
-  %}
-  {% set spark_sql = spark_sql ~ incremental_to_spark_sql %}
-
---     Insert incremental output into table
-  {% set write_to_table_sql =
-    'query_response.write.mode("append").saveAsTable("' ~ output_table ~ '")\n'
-  %}
-  {% set spark_sql = spark_sql ~ write_to_table_sql %}
+--     Generate spark sql for incremental maintenance logic
+  {% set spark_sql = coral_dbt.incremental_code_generation(coral_response, table_names, output_table) %}
```
[the remark this answers is not in the document]
Comment (AliceYeh12) [reply coral-dbt/src/main/resources/macros/coral_macros/spark/materializations/incremental_maintenance.sql:27]: 👍, will fix on PR #376 and carry over on rebase.

File: coral-dbt/src/main/resources/tests/test_incremental.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+import subprocess
+import unittest
+from jinja2 import Template
+
+SEEDS_DIR = 'seeds/'
+URL = 'http://localhost:8080/api/incremental/rewrite'
+
+class TestUtils:
+    @staticmethod
+    def construct_generate_incremental_code_test(name):
```
Comment (aastha25) [inline coral-dbt/src/main/resources/tests/test_incremental.py:15]: can you rename these to introduce more clarity? 

something like `construct_expected_incremental_scala_script` & `construct_actual_incremental_scala_script`?

File: coral-dbt/src/main/resources/tests/test_incremental.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    def test_generate_incremental_code(coral_response, table_names, output_table):
+        with open('generate_incremental_code.j2') as f:
+            template = Template(f.read())
+
+        context = {'coral_response': coral_response, 'table_names': table_names, 'output_table': output_table}
+        output = template.render(context)
+
+        return output
+
+class TestIncremental(unittest.TestCase):
```
Comment (aastha25) [inline coral-dbt/src/main/resources/tests/test_incremental.py:36]: These tests are running as a Gradle task as part of the overall build process. Can you also add steps for how to run the tests separately and from which directory they can be run?

File: coral-dbt/src/main/resources/macros/coral_macros/default/utils/generate_incremental_code.sql [hunk 8]
```diff
@@ -0,0 +1,13 @@
+/**
+ * Copyright 2023 LinkedIn Corporation. All rights reserved.
+ * Licensed under the BSD-2 Clause license.
+ * See LICENSE in the project root for license information.
+ */
+{% macro generate_incremental_code(coral_response, table_names, output_table) %}
```
Comment (aastha25) [inline coral-dbt/src/main/resources/macros/coral_macros/default/utils/generate_incremental_code.sql:6]: I feel calling it 'script' instead of 'code' reflects what this macro does. can you try and rename this to something like - 'generate_incremental_script_for_adapter / execution'?

Also, please add comments on the macros.

File: coral-dbt/src/main/resources/macros/coral_macros/default/utils/get_coral_incremental_response.sql [hunk 9]
```diff
@@ -3,11 +3,21 @@
  * Licensed under the BSD-2 Clause license.
  * See LICENSE in the project root for license information.
  */
-{% macro get_coral_incremental_response(sql, table_names) %}
-  {{ return(adapter.dispatch('get_coral_incremental_response', macro_namespace = 'coral_dbt')(sql, table_names)) }}
+{% macro get_coral_incremental_response(sql, url, table_names, sql_dialect) %}
```
Comment (aastha25) [inline coral-dbt/src/main/resources/macros/coral_macros/default/utils/get_coral_incremental_response.sql:6]: as part of the macro documentation, lets also capture the request & response data with an example here.

File: coral-dbt/src/main/resources/tests/test_incremental.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    @staticmethod
+    def test_generate_incremental_code(coral_response, table_names, output_table):
+        with open('generate_incremental_code.j2') as f:
+            template = Template(f.read())
+
+        context = {'coral_response': coral_response, 'table_names': table_names, 'output_table': output_table}
+        output = template.render(context)
+
+        return output
+
```
Comment (aastha25) [inline coral-dbt/src/main/resources/tests/test_incremental.py:35]: as we add more tests, this file might become unmanageable. 

can you evaluate if splitting it up is an option? maybe Utils + setup & teardown can be in a separate file, tests separate.

File: coral-dbt/src/main/resources/macros/coral_macros/spark/utils/generate_incremental_code.sql [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+--
+--       incremental_maintenance_sql: SELECT * FROM db_t1_delta UNION SELECT * FROM db_t2_delta
+--       underscore_delimited_table_names: ['db_t1', 'db_t2']
+--       incremental_table_names: ['db_t1_delta', 'db_t2_delta']
+  {% set incremental_maintenance_sql = coral_response['incremental_maintenance_sql'] %}
+  {% set underscore_delimited_table_names = coral_response['underscore_delimited_table_names'] %}
+  {% set incremental_table_names = coral_response['incremental_table_names'] %}
+
+--     Generate Iceberg incremental table scan code for each target table
+--     Namespace variable used to persist changes beyond loop scope
```
Comment (aastha25) [inline coral-dbt/src/main/resources/macros/coral_macros/spark/utils/generate_incremental_code.sql:23]: also add a comment here to elaborate on how the start and end snapshot ids are retrieved and that logic might change in the future.

**Review by aastha25 — COMMENTED** [review]
Great job on the PR, @AliceYeh12.No major concerns.

File: coral-dbt/src/main/resources/tests/test_incremental.py (hunk 10 above)
Comment (AliceYeh12) [reply coral-dbt/src/main/resources/tests/test_incremental.py:35]: Hmm, I feel like it makes sense to keep all of these in one file since this is all setup/teardown and utils associated with testing incremental, which can be likened to the beforeClass() and afterClass() methods and helper functions in the tests in the Java modules in Coral.

**Review by aastha25 — APPROVED** [review]
LGTM
